File: src/ui/text.ts
/**
 * 文本输入。**不用 inquirer**——它的 readline 会和自绘 selectMenu 的 raw 模式抢 stdin、收不到按键。
 * 改回现版 PS 的两套机制（同一时刻只有一套在跑，互不干扰）：
 *
 *  · readValue：raw 逐键行编辑器（与 selectMenu 同机制，已验证可用），用于 ASCII 字段（密钥/模型/effort）。
 *               密钥回显 `*`。语义对齐 Read-Value：回车空=不改、`-`=清空、Esc=取消、Ctrl+C 退出。
 *  · readText ：cooked 模式 readline（兼容中文输入法，评审④），用于中文字段（备注/自定义名/手动地址）。
 */
import { createInterface, emitKeypressEvents, type Key } from 'node:readline';

import { T } from '../i18n/index.js';

export interface ReadResult {
  changed: boolean;
  value: string;
}

/** raw 逐键读 ASCII 字段：回车空=不改、`-`=清空、其它=替换、Esc/Ctrl+C=取消。 */
export async function readValue(label: string, current: string, secret = false): Promise<ReadResult> {
  const stdin = process.stdin;
  const stdout = process.stdout;
  const cur = current === '' ? T('empty.paren') : secret ? '********' : current;
  stdout.write(`\n  ${label}  [${T('edit.current', cur)}]  ${T('edit.inputHint')}\n  > `);

  if (!stdin.isTTY) {
    const line = await readText(''); // 非交互回退
    if (line === undefined || line === '') return { changed: false, value: current };
    if (line === '-') return { changed: true, value: '' };
    return { changed: true, value: line };
  }

  emitKeypressEvents(stdin);
  const wasRaw = stdin.isRaw ?? false;
  stdin.setRawMode(true);
  stdin.resume();

  let buf = '';
  return new Promise<ReadResult>((resolve) => {
    const cleanup = (): void => {
      stdin.off('keypress', onKey);
      if (!wasRaw) stdin.setRawMode(false);
      stdin.pause();
      stdout.write('\n');
    };
    const onKey = (str: string | undefined, key: Key): void => {
      if (key?.ctrl && key.name === 'c') {
        cleanup();
        resolve({ changed: false, value: current });
        process.exit(130);
      }
      if (key?.name === 'return' || key?.name === 'enter') {
        cleanup();
        if (buf === '') resolve({ changed: false, value: current });
        else if (buf === '-') resolve({ changed: true, value: '' });
        else resolve({ changed: true, value: buf });
        return;
      }
      if (key?.name === 'escape') {
        cleanup();
        resolve({ changed: false, value: current });
        return;
      }
      if (key?.name === 'backspace') {
        if (buf.length > 0) {
          buf = buf.slice(0, -1);
          stdout.write('\b \b');
        }
        return;
      }
      // 普通字符（含粘贴的多字符），过滤控制字符
      if (str && !key?.ctrl && !key?.meta) {
        const printable = [...str].filter((c) => c >= ' ' && c !== '\x7f').join('');
        if (printable) {
          buf += printable;
          stdout.write(secret ? '*'.repeat(printable.length) : printable);
        }
      }
    };
    stdin.on('keypress', onKey);
  });
}

/** cooked 模式读一行（兼容中文输入法）；Ctrl+C/中止返回 undefined。 */
export async function readText(message: string): Promise<string | undefined> {
  const stdin = process.stdin;
  if (stdin.isTTY) stdin.setRawMode(false); // 确保 cooked，输入法才能组词
  stdin.resume();
  const rl = createInterface({ input: stdin, output: process.stdout });
  return new Promise<string | undefined>((resolve) => {
    let done = false;
    rl.on('SIGINT', () => {
      if (done) return;
      done = true;
      rl.close();
      resolve(undefined);
    });
    rl.question(message ? `${message}` : '  > ', (ans) => {
      if (done) return;
      done = true;
      rl.close();
      resolve(ans);
    });
  });
}

File: internal/tui/text.go
package tui

import (
	"os"
	"strings"

	"github.com/becomeless/cc-x/internal/i18n"
)

// ReadText 在 cooked 模式读一行（兼容中文输入法）。调用前终端应处于 cooked；ReadLine 内部会确保已 Restore。
// 返回 (line, ok)；EOF/中止且无内容时 ok=false。对应 npm 版 src/ui/text.ts 的 readText。
// 语义（空=不改、"-"=清空等）由调用方处理。
func ReadText(t *Terminal, prompt string) (string, bool) {
	return t.ReadLine(prompt)
}

// ReadValue raw 逐键读 ASCII 字段：回车空=不改、"-"=清空、其它=替换、Esc=取消、Ctrl+C 退出（130）。
// secret=true 时回显 *。对应 npm 版 readValue。
//
// 直接处理整段字节而非单键，以支持粘贴（多字符一次到达；ReadKey 单键抽象会丢掉除首字符外的内容）。
func ReadValue(t *Terminal, label, current string, secret bool) (changed bool, value string) {
	cur := current
	switch {
	case cur == "":
		cur = i18n.T("empty.paren")
	case secret:
		cur = "********"
	}
	t.Write("\n  " + label + "  [" + i18n.T("edit.current", cur) + "]  " + i18n.T("edit.inputHint") + "\n  > ")

	if !t.IsTTY() {
		return cookedFallback(t, current)
	}
	if err := t.MakeRaw(); err != nil {
		return cookedFallback(t, current)
	}

	var buf []rune
	for {
		n, err := t.In.Read(t.buf[:])
		if n == 0 && err != nil {
			t.Restore()
			t.Write("\n")
			return false, current
		}
		chunk := t.buf[:n]
		switch ParseKey(chunk).Type {
		case KeyCtrlC:
			t.Restore()
			t.Write("\n")
			os.Exit(130)
		case KeyEnter:
			t.Restore()
			t.Write("\n")
			s := string(buf)
			if s == "" {
				return false, current
			}
			if s == "-" {
				return true, ""
			}
			return true, s
		case KeyEsc:
			t.Restore()
			t.Write("\n")
			return false, current
		case KeyBackspace:
			if len(buf) > 0 {
				buf = buf[:len(buf)-1]
				t.Write("\b \b")
			}
		case KeyChar, KeyDigit:
			s := filterPrintable(chunk)
			if s != "" {
				buf = append(buf, []rune(s)...)
				if secret {
					t.Write(strings.Repeat("*", len([]rune(s))))
				} else {
					t.Write(s)
				}
			}
		}
	}
}

// cookedFallback：非 TTY / raw 失败时，cooked 读一行（语义同 ReadValue 的回车/-/替换）。
func cookedFallback(t *Terminal, current string) (bool, string) {
	line, ok := t.ReadLine("")
	if !ok || line == "" {
		return false, current
	}
	if line == "-" {
		return true, ""
	}
	return true, line
}

// filterPrintable 保留可打印字符（>=0x20 且非 DEL），过滤控制字符与转义序列残留。
func filterPrintable(b []byte) string {
	var sb strings.Builder
	for _, r := range string(b) {
		if r >= 0x20 && r != 0x7f {
			sb.WriteRune(r)
		}
	}
	return sb.String()
}
